src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_libuv.cc

```diff
@@ -0,0 +1,165 @@
+  }
+
+  void RegisterForOnWriteableLocked(grpc_closure* write_closure) override {
+    GPR_ASSERT(write_closure_ == nullptr);
+    GPR_ASSERT((poll_events_ & UV_WRITABLE) == 0);
+    write_closure_ = write_closure;
+    poll_events_ |= UV_WRITABLE;
+    uv_poll_start(handle_, poll_events_, ares_uv_poll_cb);
+  }
+
+  bool IsFdStillReadableLocked() override {
+    /* uv_poll_t is based on poll, which is level triggered. So, if cares
+     * leaves some data unread, the event will trigger again. */
+    return false;
+  }
+
+  void ShutdownLocked(grpc_error* error) override {
+    grpc_core::ExecCtx exec_ctx;
+    uv_poll_stop(handle_);
+    uv_close(reinterpret_cast<uv_handle_t*>(handle_), ares_uv_poll_close_cb);
+    if (read_closure_) {
+      GRPC_CLOSURE_SCHED(read_closure_, GRPC_ERROR_CANCELLED);
+    }
+    if (write_closure_) {
+      GRPC_CLOSURE_SCHED(write_closure_, GRPC_ERROR_CANCELLED);
+    }
+  }
+
+  ares_socket_t GetWrappedAresSocketLocked() override { return as_; }
+
+  const char* GetName() override { return name_; }
+
+  char* name_;
+  ares_socket_t as_;
+  uv_poll_t* handle_;
+  grpc_closure* read_closure_ = nullptr;
+  grpc_closure* write_closure_ = nullptr;
+  int poll_events_ = 0;
+  grpc_combiner* combiner_;
+};
+
+struct AresUvPollCbArg {
+  AresUvPollCbArg(uv_poll_t* handle, int status, int events)
+      : handle(handle), status(status), events(events) {}
+
+  uv_poll_t* handle;
+  int status;
+  int events;
+};
+
+static void inner_callback(void* arg, grpc_error* error) {
+  AresUvPollCbArg* arg_struct = reinterpret_cast<AresUvPollCbArg*>(arg);
+  uv_poll_t* handle = arg_struct->handle;
+  int status = arg_struct->status;
+  int events = arg_struct->events;
+  GrpcPolledFdLibuv* polled_fd =
+      reinterpret_cast<GrpcPolledFdLibuv*>(handle->data);
+  if (status < 0) {
+    error = GRPC_ERROR_CREATE_FROM_STATIC_STRING("cares polling error");
+    error =
+        grpc_error_set_str(error, GRPC_ERROR_STR_OS_ERROR,
+                           grpc_slice_from_static_string(uv_strerror(status)));
+  }
+  if ((events & UV_READABLE) && polled_fd->read_closure_) {
+    GRPC_CLOSURE_SCHED(polled_fd->read_closure_, error);
+    polled_fd->read_closure_ = nullptr;
+    polled_fd->poll_events_ &= ~UV_READABLE;
+  }
+  if ((events & UV_WRITABLE) && polled_fd->write_closure_) {
+    GRPC_CLOSURE_SCHED(polled_fd->write_closure_, error);
+    polled_fd->write_closure_ = nullptr;
+    polled_fd->poll_events_ &= ~UV_WRITABLE;
+  }
+  uv_poll_start(handle, polled_fd->poll_events_, ares_uv_poll_cb);
+  Delete(arg_struct);
+}
+
+void ares_uv_poll_cb(uv_poll_t* handle, int status, int events) {
+  grpc_core::ExecCtx exec_ctx;
+  GrpcPolledFdLibuv* polled_fd =
+      reinterpret_cast<GrpcPolledFdLibuv*>(handle->data);
+  AresUvPollCbArg* arg = New<AresUvPollCbArg>(handle, status, events);
+  GRPC_CLOSURE_SCHED(
+      GRPC_CLOSURE_CREATE(inner_callback, arg,
+                          grpc_combiner_scheduler(polled_fd->combiner_)),
```

Comment:
nit: even though this `GrpcPolledFd` lives within an ev driver lifetime (thus the combiner will be kept alive), can we ref and unref this combiner since it's being used now?